node/lib/util/merge_common.js

```diff
@@ -0,0 +1,317 @@
+ */
+class MergeContext {
+    /**
+    * @param {NodeGit.Repository}      repo
+    * @param {NodeGit.Commit|null}     ourCommit
+    * @param {NodeGit.Commit}          theirCommit
+    * @param {MergeCommon.MODE}        mode
+    * @param {Open.WDTYPE}             wdType
+    * @param {String|null}             commitMessage
+    * @param {() -> Promise(String)}   editMessage
+    */
+    constructor(metaRepo,
+                ourCommit,
+                theirCommit,
+                mode,
+                wdType,
+                commitMessage,
+                editMessage) {
+        assert.instanceOf(metaRepo, NodeGit.Repository);
+        if (null !== ourCommit) {
+            assert.instanceOf(ourCommit, NodeGit.Commit);
+        }
+        assert.instanceOf(theirCommit, NodeGit.Commit);
+        assert.isNumber(mode);
+        assert.isNumber(wdType);
+        if (null !== commitMessage) {
+            assert.isString(commitMessage);
+        }
+        assert.isFunction(editMessage);
+        this.d_metaRepo = metaRepo;
+        this.d_ourCommit = ourCommit;
+        this.d_theirCommit = theirCommit;
+        this.d_mode = mode;
+        this.d_wdType = wdType;
+        this.d_commitMessage = commitMessage;
+        this.d_editMessage = editMessage;
+        this.d_opener = new Open.Opener(metaRepo, null);
+        this.d_changeIndex = null;
+        this.d_changes = null;
+        this.d_conflictsMessage = "";
+    }
+
+    /**
+     * @property {Boolean} forceBare if working directory is disabled
+     */
+    get forceBare() {
+        return Open.WDTYPE.FORCE_BARE === this.d_wdType;
+    }
+
+    /**
+     * @property {NodeGit.Repository}
+     */
+    get metaRepo() {
+        return this.d_metaRepo;
+    }
+
+    /**
+     * @property {Opener}
+     */
+    get opener() {
+        return this.d_opener;
+    }
+
+    /**
+     * @property {NodeGit.Commit}
+     */
+    get theirCommit() {
+        return this.d_theirCommit;
+    }
+
+    /**
+     * @property {Open.WDTYPE}
+     */
+    get wdType() {
+        return this.d_wdType;
+    }
+
+    /**
+     * @property {MODE}
+     */
+    get mode() {
+        return this.d_mode;
+    }
+}
+
+/**
+ * @async
+ * @return {Object} return from sub name to `SubmoduleChange`
+ * @return {Object} return.simpleChanges from sub name to `Submodule`
+ * @return {Object} return.changes from sub name to `Submodule`
+ * @return {Object} return.conflicts from sub name to `Conflict`
+ */
+MergeContext.prototype.getChanges = co.wrap(function *(){
+    if (null === this.d_changes) {
+        this.d_changes =
+            yield CherryPickUtil.computeChangesBetweenTwoCommits(
+                this.d_metaRepo,
+                yield this.getChangeIndex(),
+                yield this.getOurCommit(),
+                this.d_theirCommit);
+    }
+    return this.d_changes;
+});
+
+/** 
+ * @async
+ * @return {NodeGit.Commit} return left side merge commit
```

Comment:
we need to determine if `ourCommit` comes from direct user input or from "HEAD", we can either determine that at the very beginning or on the fly when needed. I choose the later case because I think being lazy is always good.

Reply:
It's best to have separation of concerns.  One piece of code is responsible for figuring out what's going to be merged.  Another is responsible for doing the merge.

Reply:
yes, `getOurCommit` does exactly that. It is a helper that determines check if left side commit is given, and if not given use the `HEAD` if the we are not merging in a bare repo. 

I think this logic is unavoidable, the question is where we put this logic. IMO, the context object is a good place because it manages the objects that are used in the merge and `ourCommit` is one of the managed objects.